More Composition analysis methods

mortonjt opened the pull request:
This PR does the following 

* Adds 4 new methods `inner`, `ilr`, `ilr_inv` and `clr_inv`
* Clarifies some mathematical details from previous PR
* Cleans up some of the mathjax equations 

This will be the last major PR realized for visualizations.  Like I said before, these methods will complement phylogenetic visualization very nicely.  In addition, these methods are a stepping stone for compositional statistical analysis.

Jorge-C commented:
Cool! There's a test that doesn't pass in all builds (seems to be an accuracy thing) and a doctest too.

mortonjt commented:
I'm right now benchmarking these methods, so hold on a bit so I can more rigorously test some of these methods.

mortonjt commented:
Looks like the last run time out ... Mind if we could restart the Travis build?

Once this passes, @Jorge-C , would you mind to take a look over this PR once you have time?
I'm confident that all of the bugs have been weeded out.

ebolyen commented:
@mortonjt @Jorge-C Besides, the merge conflict, what is the status of this PR?

Jorge-C commented:
I believe it was ready when I removed the do not merge tag, not sure if Jamie wanted to add anything else here.

ElDeveloper commented:
Just FYI, @mortonjt is out of town until Monday.

On (Aug-26-15|15:18), Jorge Cañardo Alastuey wrote:
> I believe it was ready when I removed the do not merge tag, not sure if Jamie wanted to add anything else here. 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/scikit-bio/pull/892#issuecomment-135194241

mortonjt commented:
Alright.  I fixed a potential issue with multiplicative replacement.

Basically, it was possible for this replacement strategy to generate negative proportions with the default parameters before.  Just fixed the default parameter in this last commit.

@Jorge-C , do you mind if you could take a look over this?

Jorge-C commented:
In the new description of `delta`:
> :math:`\delta = \frac{1}{N M}` where :math:`N`
is the number of components and :math:`M` is the
number of individuals found in the largest sample

but from looking at the code, M seems to be the maximum number of zero proportions in a sample. Are both equivalent?

mortonjt commented:
Not quite.  M is the largest sequencing depth.

Here's the details why I changed the default parameter for this method.

In an edge case, there are N-1 zeros in the sample, and there is only 1 read in the remaining taxa, it is impossible for the replaced zeros to be larger than that taxa with the current default parameter.  In addition, if `(N-1) * delta > 1`, then the scaling factor will become negative and that remaining taxa is now a negative proportion.

Jorge-C commented:
Ok, I just saw why docs and code seemed inconsistent to me, although I don't understand how it happens  :confused:

When looking at the changes from your [new commit](https://github.com/mortonjt/scikit-bio/commit/25875c6dc4b960ae2a2c3f32513fe0a8f1e36021#diff-704d4acf77c05c8963c80a29372d7343L189), even before the merge from master, delta is computed in a different way. However, when checking the full diff between your PR and master, the way it's computed is consistent with the new docs but the only commit in between is a merge from master that doesn't modify `compositions.py` (?)

I'll add a minor comment inline, but I'm also interested to know how did that line change after `25875c6` :smile:

Jorge-C commented on skbio/stats/composition.py:
```diff
@@ -150,8 +158,9 @@ def multiplicative_replacement(mat, delta=None):
-       is the number of components
+       :math:`\delta = \frac{1}{N M}` where :math:`N`
+       is the number of components and :math:`M` is the
+       number of individuals found in the largest sample
```
This description doesn't seem too clear in the module's context, maybe it should use composition instead of sample, and expand on what number of individuals means in a matrix of proportions.

mortonjt replied:
Oh ....

Crap.  That was poor oversight.  I remember why I used `delta=N^2` before.  This was appropriate for proportions.  The `delta=1/NM` is appropriate for counts.  I'm going to revert ...

mortonjt commented:
That was my fault.  It should have been `mat.sum(axis=-1).max()`, not `tot.max()`

I know that the math is a bit dense.  If you'd like, we can use Google Hangouts and maybe Groupboard to sketch out some of the proofs behind these methods.

Jorge-C commented:
Thanks! The math is fine, this is just a git question :) `25875c6` uses `tot.max()`, the following commit `d00db53` doesn't modify that line, but somehow the combination no longer uses `tot.max()`. Maybe it's a github glitch?

mortonjt commented:
I actually corrected this in `d00db53` right before the merge.  So the `delta` in `25875c6` is wrong.

Jorge-C commented:
Ah, I see. It could be useful to move that change from `d00db53` to `25875c6` (and force push, new commits will be created) so that the problem is fixed in the commit that says so in the log instead of in the merge, but that's some added git magic just for tidiness, so it's your call.

mortonjt commented:
I'm probably going to squash these commits, so that problem should disappear.  Sound good?

Jorge-C commented:
:+1:
On Sep 2, 2015 12:05 PM, "mortonjt" <notifications@github.com> wrote:

> I'm probably going to squash these commits, so that problem should
> disappear. Sound good?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/892#issuecomment-137193402>.
>

mortonjt commented:
I'll start squashing right before merge.

Is that the only problem?

Jorge-C commented:
The minor doc inline comment too, maybe a native speaker could double check it?

mortonjt commented:
Alright, removed that docstring and default parameter.  It was back to its original status.

Also added those experimental decorators.  Mind you could take another look @Jorge-C?

Jorge-C commented:
:+1: do you mind squashing if possible?

mortonjt commented:
Consider it squashed :)

Jorge-C commented:
Wonderful, thank you Jamie!

mortonjt commented:
Thanks!!
